extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/processors/join/CloseableIterator.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.rya.streams.kafka.processors.join;
+
+import java.util.Iterator;
+
+import edu.umd.cs.findbugs.annotations.DefaultAnnotation;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * An {@link Iterator} that is also {@link AutoCloseable}.
+ *
+ * @param <T> - The type of elements that will be iterated over.
+ */
+@DefaultAnnotation(NonNull.class)
+public interface CloseableIterator<T> extends Iterator<T>, AutoCloseable { }
```

Comment:
Do we really need another CloseableIterator interface?

Reply:
Yes.

info.aduna.iteration.CloseableIteration does not extend AutoCloseable, so I can't use try with resources blocks on it.

org.eclipse.rdf4j.common.iteration.CloseableIteration is part of rdf4j. We use an sesame right now, so I can not access that class. But even if I could, it uses the info.aduna.iteration.Iteration interface to perform iteration, which is not for each loop compatible.

This class only uses Java interfaces. It allows me to use both the try with resources and the for each loop features of Java. If you know of another class in our project that uses only the Java interfaces, I'll gladly swap over.

Reply:
I thought that you implemented another instance of CloseableIterator for similar reasons when you were doing the Fluo work.  Did that get deleted?

Reply:
Somebody deleted that.

Reply:
Or at least I thought they did, let me check again.

Reply:
I found a static one inside of PrecomputedJoinStorage. Where do you want me to move it to? That's in the secondary indexing project, which I do not depend on. I could move it to the api project.

Reply:
Moved to api and updated all of the references.